rust/capture/src/utils.rs

```diff
@@ -36,3 +36,148 @@ pub fn uuid_v7() -> Uuid {
 
     encode_unix_timestamp_millis(now_millis, &bytes)
 }
+
+pub fn replace_invalid_hex_escape_strings(
+    json_str: String,
+) -> Result<String, std::string::FromUtf8Error> {
+    // Consume the String and get its bytes
+    let mut bytes = json_str.into_bytes();
+    let len = bytes.len();
+    let mut i = 0;
+
+    while i < len {
+        if bytes[i] == b'\\' && i + 1 < len && bytes[i + 1] == b'u' {
+            // Check if there are enough bytes for a Unicode escape sequence
+            if i + 6 <= len {
+                // Extract the four escape sequence bytes
+                let mut code_point_bytes: [u8; 4] =
+                    [bytes[i + 2], bytes[i + 3], bytes[i + 4], bytes[i + 5]];
+
+                // Convert the bytes to a string, then parse it into a u16 to check if it's a valid escape sequence
+                if let Ok(Ok(num)) =
+                    std::str::from_utf8(&code_point_bytes).map(|s| u16::from_str_radix(s, 16))
+                {
+                    if (0xD800..=0xDBFF).contains(&num) {
+                        // High surrogate without a following low surrogate
+                        if !is_next_low_surrogate(&bytes, i + 6) {
+                            // Replace with 'FFFD' (Unicode replacement character)
+                            code_point_bytes.copy_from_slice(b"FFFD");
```

Comment:
when knowing so little of the language we can't meaningfully review we instead become nit-picky

high surrogate, low surrogate, and replacement character are repeated multple times so we could extract them and give them names

but i can live with not doing that

Reply:
Agreed, done - fwiw, my goal here is that you walk away feeling like "yeah, I could muck around a bit in here (with help) if I needed to" - if you've got questions, ask em (or feel free to sync on it)

Reply:
Yes, is awesome - thank you!

I'm pleased that for anything in future I can at least write a failing test now so I can express my problem clearly if nothing else, 

and since i'm set up to run tests and hit the debugger i can at least start _trying_